[quant][graphmode] Skip quantizing input and output in matched module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33650 [not4land] debugging
* #33533 [not4land] Temporary workaround to get quantized ops
* #33895 [quant][graphmode] insert prepack/unpack for aten::linear
* #33854 [quant][graphmode] Quantization pattern for aten::linear
* #33532 [quant][graphmode] Add quantization pattern for quantized::add_relu
* #33279 [quant][graphmode] Add quantized conv2d-relu fusion pattern
* #33927 [quant][graphmode] Add Finalize function that inlines graph and produce quantized ops
* #33853 [quant][graphmode] Swap quantized functional linear with aten::linear
* #33481 [quant][graphmode] Handling ops doesn't require observation in insertObservers
* #32816 [quant][graphmode] Add add_relu pattern in skip values
* **#32814 [quant][graphmode] Skip quantizing input and output in matched module**

Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208785](https://our.internmc.facebook.com/intern/diff/D20208785)

jerryzh168 commented:
description needs to be updated

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 52fcfb38:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=52fcfb38e2edf72180673953a76f67a5f7400933).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4377580) caffe2_onnx_py2_gcc5_ubuntu16_04_build (1/1)
**Step:** "Set Up CI Environment After attach_workspace" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=93000195) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=93000195)) :snowflake:
<details>
<summary>
<code>gpg: no valid OpenPGP data found.</code>
</summary>

```
+ curl -s -L --retry 3 https://nvidia.github.io/nvidia-docker/gpgkey 
+ sudo apt-key add - 
gpg: no valid OpenPGP data found. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

dreiss commented on test/test_jit.py:
```diff
@@ -1515,7 +1510,7 @@ def test_foldbn_trivial_nobias(self):
             def __init__(self):
                 super(TestModule, self).__init__()
-                self.conv = torch.nn.Conv2d(1, 20, 5, 1, bias=False)
+            self.conv = torch.nn.Conv2d(1, 20, 5, 1, bias=False)
```
This looks like a typo.

dreiss commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -413,30 +441,40 @@ class InsertObserversHelper {
-    return (%r) )");
-  const PatternInfo matmul_add = PatternInfo::parse_from_str(R"(
+    %first_module = match::module[name="Conv2d"](%self)
+    %first_output = prim::CallMethod[name="forward"](%first_module, %input)
```
Does this handle the case where you call the module, rather than calling forward directly?

jerryzh168 replied:
I think call the module is same as calling forward method: https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L540

dreiss commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -649,30 +687,151 @@ void InsertObserversHelper::insertObserverFor(
+      Value* first_output) {
+  Value* self = graph.inputs()[0];
+  std::vector<std::string> first_module_path = getModuleAccessPath(first_module, self);
+  TORCH_CHECK(first_module_path.size() != 0, "Don't support skipping "
```
I would just join these strings and let clang-fmt figure it out.

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/test_jit.py:
```diff
@@ -1127,35 +1126,31 @@ def __init__(self):
+        assert len(attrs_with_prefix(m.conv, '_observer_')) == 1, \
+            'Expected to have 1 observer submodule'
+        # observer for output of relu
+        assert len(attrs_with_prefix(m.relu, '_observer_')) == 1, \
```
Should this be zero?

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208785](https://our.internmc.facebook.com/intern/diff/D20208785)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208785](https://our.internmc.facebook.com/intern/diff/D20208785)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Skip quantizing input and output in matched module"


Summary:
We skip quantization for the intermediate values for patterns like `Conv - ReLU`,
but currently we didn't skip quantizing the input/output of the graphs of matched modules,
since we now changed the way we add observers, this also needs to be updated.

Test Plan:
python test/test_jit.py -- 'TestJit.test_insert_observers_skip_values'
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208785](https://our.internmc.facebook.com/intern/diff/D20208785)

[ghstack-poisoned]

ljk53 commented:
@jerryzh168 do you think this might be related to the master CI failure: https://app.circleci.com/jobs/github/pytorch/pytorch/4684269 ?

This PR stacked with my PR #34055 so it's not obvious which one is the cause, but mine shouldn't affect test_jit :) cc: @mrshenli

jerryzh168 commented:
This PR is not changing jit and `test_script_pack_padded_sequence` is not a quantization test.
Is this failing consistently?
